42.md

```diff
@@ -19,35 +19,51 @@ A relay may want to require clients to authenticate to access restricted resourc
     before clients can query for that kind.
   - A relay may limit subscriptions of any kind to paying users or users whitelisted through any other means, and require authentication.
 
-## Protocol flow
+## Definitions
+
+This NIP defines a new message, `AUTH`, which relays can send when they support authentication and clients can send to relays when they want
+to authenticate. When sent by relays, the message is of the following form:
+
+```
+["AUTH", <challenge-string>]
+```
```

Comment:
We can add that later if necessary.

Reply:
So I would implement the challenge upon a socket being established. If the client somehow lost the nonce, it could always re-connect.

The nip is vague about when the relay might challenge indeed. Also a client might want to switch authentication, so while authed as Alice it might want to switch to auth as Bob now, so both for un-authed and authed clients there is a need to request a new nonce. Semisol's empty AUTH message triggering this has my support.

Reply:
I agree and it's so simple to implement we should try to have it from the start.

Reply:
I think it is unnecessary bloat. How does that work on the relay side? The relay is emitting events for identity A at a given websocket, then that websocket changes to identity B, now the relay has to shove that inside the code it might have for handling the subscriptions, maybe cancel a subscription that was valid?

Reply:
The client should just open a new websocket if it wants a new identity -- which is also an edge case I see no use for and we shouldn't be complicating this for 99.9% of the cases.

Reply:
> The nip is vague about when the relay might challenge indeed

This is because it is irrelevant. All clients would just set up a listener like `when("AUTH") => send("AUTH")` and be done with it. Most relays will send the `AUTH` challenge on connect, but why specify that if it changes nothing on the code side (the client has to wait anyway)?

Also, if you think the client sending an empty `AUTH` request fixes it, consider that the relay may still take 5 minutes to reply to that with a challenge -- or may never reply, so really nothing is changed.